finetuner/tuner/base.py

```diff
@@ -1,112 +1,138 @@
 import abc
 import warnings
-from typing import (
-    Generator,
-    Optional,
-    Union,
-    Tuple,
-    List,
-    Dict,
-)
-
-from ..helper import AnyDNN, AnyTensor, AnyDataLoader, AnyOptimizer, DocumentArrayLike
+from typing import Callable, Dict, Generic, List, Optional, Tuple, TypeVar, Union
+
+from .dataset import ClassDataset, SessionDataset
+from .dataset.samplers import RandomClassBatchSampler, SessionBatchSampler
 from .summary import Summary
+from ..helper import AnyDataLoader, AnyDNN, AnyOptimizer, AnyTensor, DocumentSequence
+
+
+LabelType = TypeVar('LabelType')
+
+
+class BaseMiner(abc.ABC, Generic[AnyTensor, LabelType]):
+    @abc.abstractmethod
+    def mine(self, labels: LabelType, distances: AnyTensor) -> Tuple[AnyTensor, ...]:
+        """Generate mined tuples from labels and item distances.
+
+        :param labels: labels of each item
+        :param distances: A tensor matrix of pairwise distance between each two item
```

Comment:
now sampler give distances rather than embeddings?

Reply:
Yes, passing embeddings was inefficient - if there were a million triplets, it would have to compute million distances, instead of taking a million numbers from distance matrix. In almost all cases there are more triplets/pairs than elements in distance matrix, so this is much more efficient